File: python/CLRS/chapter2/merge_sort.py
# Merge sort
# Avg time complexity: O(nlog n)
# Best time comlexity: O(nlog n)
# Worst time complexity: O(nlog n)
# Space complexity: O(n) if data structure is array. Constant space if data
# structure is linked list.
#
# The key operation of the merge sort algoirthm is the merging of two sorted
# sequences in the "combine" step in divide and conquer approach.
# We break down the array structure in sub arrays and sort them. In its
# simplist sub-array form, each sub-array has only 1 item, which is by itself
# sorted. Then join 2 sub-arrays into a 'greater' sub-array by retrieving the
# elements in ascending order from each of the 2 sub-array. Repeat this until
# the there are no sub-arrays left.


def merge(li, start, mid, end):
    left_li = []
    left_li.extend(li[start:mid+1])
    right_li = []
    right_li.extend(li[mid+1:end+1])
    left_li.append(2147483647)
    right_li.append(2147483647)
    i = 0
    j = 0
    for k in range(start, end+1):
        if left_li[i] <= right_li[j]:
            li[k] = left_li[i]
            i += 1
        else:
            li[k] = right_li[j]
            j += 1


def merge_sort(li, start, end):
    if start < end:
        mid = (start + end) / 2
        merge_sort(li, start, mid)
        merge_sort(li, mid+1, end)
        merge(li, start, mid, end)


if __name__ == "__main__":
    test_list1 = [5, 2, 4, 6, 1, 3, 0]
    merge_sort(test_list1, 0, len(test_list1)-1)
    print test_list1

File: php/sorting/merge_sort.php
<?php

/**
 * Merge sort.
 * Typical divide and conquer approach.
 * Recursively divide the array to half the size until there is only 1 element
 * left in each subarray. Then merge the subarrays in ascending/descending order
 * until the original array is merged back in ascending/descending order.
 *
 * Worst time complexity: O(nlgn)
 * Worst space complexity: O(n)
 */

$input1 = [14, 33, 27, 10, 35, 19, 42, 44];
$input2 = [64, 25, 12, 22, 11];

function mergesort($input) {
    if (sizeof($input) <= 1) {
        return $input;
    }

    $mid = (int) (sizeof($input) / 2);
    $left = array_slice($input, 0, $mid);
    $right = array_slice($input, $mid);
    $left = mergesort($left);
    $right = mergesort($right);
    return merge($left, $right);
}
function merge($left, $right) {
    $result = [];

    while (!empty($left) && !empty($right)) {
        if ($left[0] < $right[0]) {
            $result[] = array_shift($left);
        } else {
            $result[] = array_shift($right);
        }
    }
    if (!empty($left)) {
        $result = array_merge($result, $left);
    }
    if (!empty($right)) {
        $result = array_merge($result, $right);
    }
    return $result;
}

print_r(mergesort($input1));